lib/Dialect/Arc/Transforms/FindInitialVectors.cpp

```diff
@@ -0,0 +1,222 @@
+      if (auto stateOp = dyn_cast<StateOp>(op)) {
+        if (stateOp.getLatency() > 0)
+          item.userIt = op->user_end();
+      } else if (auto writeOp = dyn_cast<MemoryWritePortOp>(op)) {
+        item.userIt = op->user_end();
+      }
+      if (item.userIt == op->user_end()) {
+        opRanks.insert({op, item.rank});
+        worklist.pop_back();
+        continue;
+      }
+      if (auto rankIt = opRanks.find(*item.userIt); rankIt != opRanks.end()) {
+        item.rank = std::max(item.rank, rankIt->second + 1);
+        ++item.userIt;
+        continue;
+      }
+      if (!worklist.insert({*item.userIt, WorklistItem(*item.userIt)}).second)
+        return op->emitError("dependency cycle");
+    }
+  }
+  return success();
+}
+
+namespace {
+template <typename T>
+LogicalResult modifyKeyIfPossible(const T &container, std::string &key) {
+  if(container.empty())
+    return success();
+
+  if constexpr (std::is_same_v<T, StringAttr>) {
+    // The attributes' order doesn't matter so we need to sort the attributes
+    // in some order and the concatinate them
+    std::vector tempAttributeHolder(container.begin(), container.end());
+    std::sort(tempAttributeHolder.begin(), tempAttributeHolder.end());
+    key += std::accumulate(tempAttributeHolder.begin(), tempAttributeHolder.end(), "");
+    return success();
+  }
+
+  // OperandTypes or ResultTypes
+  for (const auto &elem : container) {
+    std::string str;
+    llvm::raw_string_ostream os(str);
+    elem.print(os);
+    key += str;
+  }
+  return success();
+}
+
+std::string computeKey(Operation *op) {
+  // The key = concat(op_rank, op_name, op_operands_types, op_result_types, op_attrs)
+  std::string key = op->getName().getStringRef().str();
+  if (failed(modifyKeyIfPossible(op->getOperandTypes(), key)) ||
+      failed(modifyKeyIfPossible(op->getResultTypes(), key)) || 
+      failed(modifyKeyIfPossible(op->getRegisteredInfo()->getAttributeNames(), key)))
+    assert(false && "Cannot compute the op key");
+
+  return key;
+}
+
+struct Vectorizer {
+  LogicalResult collectSeeds(Block *block) {
+    if (failed(order.compute(block)))
+      return failure();
+
+    for(auto &[op, rank] : order.opRanks)
+      candidates[std::to_string(rank) + computeKey(op)].push_back(op);
+
+    return success();
+  }
+
+  LogicalResult vectorize();
+  // Store Isomorphic ops together
+  std::unordered_map<std::string, SmallVector<Operation *>> candidates;
+  TopologicalOrder order;
+};
+} // namespace
+
+// When calling this function we assume that we have the candidate groups of
+// isomorphic ops so we need to feed them to the `VectorizeOp`
+LogicalResult Vectorizer::vectorize() {
+  // Unachievable?! just in case!
+  if (candidates.empty())
+    return success();
+
+  // Iterate over every group of isomorphic ops  
+  for (const auto &[key, ops] : candidates) {
+    // If the group has only one scalar then it doesn't worth vectorizing
+    if (ops.size() == 1)
+      continue;
```

Comment:
I think `arc.vectorize` currently only supports a single result in its body region. (MLIR doesn't support VariadicOfVariadic results yet.) It would make sense to skip any ops that have more than one result here, until we have extended `arc.vectorize` to support multi-result ops.